// react-m/lib/button/flatbutton.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _rippleJsx = require('../ripple.js');

var _rippleJsx2 = _interopRequireDefault(_rippleJsx);

var _reactVendorPrefixes = require('react-vendor-prefixes');

var _reactVendorPrefixes2 = _interopRequireDefault(_reactVendorPrefixes);

var FlatButton = (function (_React$Component) {
  _inherits(FlatButton, _React$Component);

  function FlatButton() {
    _classCallCheck(this, FlatButton);

    _get(Object.getPrototypeOf(FlatButton.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(FlatButton, [{
    key: 'render',
    value: function render() {
      var _this = this;

      var _props = this.props;
      var type = _props.type;
      var enabled = _props.enabled;
      var ripple = _props.ripple;
      var color = _props.color;
      var rippleColor = _props.rippleColor;
      var children = _props.children;

      if (!enabled) {
        return _react2['default'].createElement(
          'button',
          { style: _extends({}, styles.button, styles.disabled), disabled: true },
          children
        );
      }

      var styleButton = _extends({}, styles.button);
      styleButton.color = color;

      var markupRipple;
      if (ripple) markupRipple = _react2['default'].createElement(_rippleJsx2['default'], { center: false, color: rippleColor, container: styles.ripple });

      return _react2['default'].createElement(
        'button',
        {
          style: styleButton,
          className: 'reactmd-button-flatbutton',
          onClick: this.onClick.bind(this),
          onMouseLeave: this.onMouseLeave.bind(this),
          ref: function (e) {
            return _this.button = e;
          }
        },
        children,
        markupRipple
      );
    }
  }, {
    key: 'onClick',
    value: function onClick() {
      if (this.props.onClick) this.props.onClick();
    }
  }, {
    key: 'onMouseLeave',
    value: function onMouseLeave() {
      this.button.blur();
    }
  }], [{
    key: 'propTypes',
    value: {
      onClick: _react2['default'].PropTypes.func,
      color: _react2['default'].PropTypes.string,
      enabled: _react2['default'].PropTypes.bool,
      ripple: _react2['default'].PropTypes.bool,
      rippleColor: _react2['default'].PropTypes.string
    },
    enumerable: true
  }, {
    key: 'defaultProps',
    value: {
      enabled: true,
      ripple: false,
      color: '#000',
      rippleColor: '#000'
    },
    enumerable: true
  }]);

  return FlatButton;
})(_react2['default'].Component);

exports.FlatButton = FlatButton;
exports['default'] = FlatButton;

var styles = _reactVendorPrefixes2['default'].prefix({
  button: {
    //color: '#000',

    fontFamily: 'Roboto,Helvetica,Arial,sans-serif',
    fontSize: 14,
    fontWeight: 500,
    textTransform: 'uppercase',
    letterSpacing: 0,

    height: 36,
    minWidth: 64,
    lineHeight: '36px',
    padding: '0 16px',

    border: 'none',
    borderRadius: 2,
    position: 'relative',
    display: 'inline-block',
    overflow: 'hidden',
    willChange: 'box-shadow,transform',
    transition: 'box-shadow .2s cubic-bezier(.4,0,1,1),background-color .2s cubic-bezier(.4,0,.2,1),color .2s cubic-bezier(.4,0,.2,1)',
    outline: 'none',
    cursor: 'pointer',
    textDecoration: 'none',
    textAlign: 'center',
    verticalAlign: 'middle',
    WebkitTapHighlightColor: 'rgba(255,255,255,0)'
  },

  disabled: {
    color: 'rgba(0,0,0,.26)',
    cursor: 'default',
    backgroundColor: 'transparent'
  },

  ripple: {
    top: 0,
    left: 0,
    height: '100%',
    width: '100%',
    borderRadius: 2
  }
});
